vllm/v1/core/kv_cache_manager.py

```diff
@@ -36,73 +169,413 @@ def __init__(
+        block_hashes = self.hash_prompt_tokens(request.prompt_token_ids)
+
+        for block_hash in block_hashes:
+            # block_hashes is a chain of block hashes. If a block hash is not
+            # in the cached_block_hash_to_id, the following block hashes are
+            # not computed yet for sure.
+            if cached_block := self._get_cached_block(block_hash):
+                computed_blocks.append(cached_block)
+            else:
+                break
+
+        return computed_blocks
 
     def append_slots(
         self,
         request: Request,
         num_tokens: int,
-    ) -> Optional[List[int]]:
+    ) -> Optional[List[KVCacheBlock]]:
+        """Append slots to the block table of the request.
+        We first append slots to already allocated blocks. If the allocated
+        blocks are not enough, we allocate new blocks.
+
+        Args:
+            request: The request to append slots.
+            num_tokens: The number of tokens to append.
+        
+        Returns:
+            A list of new blocks if new blocks are allocated, or None
+            if new blocks are required but cannot be allocated.
+        """
         num_required_blocks = cdiv(request.num_computed_tokens + num_tokens,
                                    self.block_size)
-        req_block_ids = self.req_to_block_ids[request.request_id]
-        if num_required_blocks <= len(req_block_ids):
-            # No new block is needed.
-            return []
+        req_blocks = self.req_to_blocks[request.request_id]
 
-        num_new_blocks = num_required_blocks - len(req_block_ids)
-        num_free_blocks = len(self.free_block_ids)
-        if num_new_blocks > num_free_blocks:
-            # Cannot allocate new blocks.
+        num_new_blocks = num_required_blocks - len(req_blocks)
+        if num_new_blocks > self.free_block_queue.num_free_blocks:
+            # Need to allocate new blocks due to insufficient pre-allocated
+            # slots, but we cannot allocate new blocks due to the limit.
             return None
 
-        # Allocate new blocks.
+        # When caching is enabled, assign token IDs to already allocated blocks.
+        new_token_ids = None
+        parent_block = None
+        if self.enable_caching:
+            # Figure out the token IDs to add to the blocks.
+            if request.num_computed_tokens < request.num_prompt_tokens:
+                # (Chunked) Prefill.
+                new_token_ids = request.prompt_token_ids[
+                    request.num_computed_tokens:request.num_computed_tokens +
+                    num_tokens]
+            else:
+                # Decode.
+                num_computed_output_tokens = (request.num_computed_tokens -
+                                              request.num_prompt_tokens)
+                new_token_ids = request.output_token_ids[
+                    num_computed_output_tokens:num_computed_output_tokens +
+                    num_tokens]
```

Comment:
Can we just do
```python
num_token_ids = request.all_token_ids[num_computed_tokens:num_computed_tokens + num_tokens]
```
And add the `all_token_ids` attribute to `request`?

Reply:
I'm asking because these two cases do not always fall into chunked prefill and decode.

For example, when re-computation happens, there could be a case like `num_prompt_tokens < num_computed_tokens < num_prompt_tokens + num_output_tokens`.

In general, I do really want to avoid the use of terms `prefill` and `decode` in V1, since these concepts do not exist at all in V1.

Reply:
I implemented in this way exactly because `Request` doesn't have `all_tokens` so I'm fine with adding this. One way I could do is adding a property to `Request`:

```python
@property
def all_token_ids(self):
    return self.prompt_token_ids + self.output_token_ids
```

But I'm not sure if the overhead of this method is acceptable.

Reply:
Good point. I agree that the overhead can be concerning. Can we have the `all_token_ids` attribute and always update it along with `output_token_ids`?